### PR TITLE
Allow to check against haveibeenpwned.com password list

### LINKED ISSUE #58: HIBP implementation (check for previous exploited passwords)
Alongside the settings to require certain password characteristics (eg. length and such) it would be cool to implement this feature that checks if the users password previously has been exploited:

https://haveibeenpwned.com/API/v2#PwnedPasswords

Original issue: [#8494](https://github.com/nextcloud/server/issues/8494)

### PR DESCRIPTION
**Pull request by rullzer**
fixes #58

Allow to check against the password list of haveibeenpwned.com

Used the range method (obviously) as described in: https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange

Before anyone asks shipping the file is very non practical see the size of https://haveibeenpwned.com/Passwords ;)

Todo:
- [ ] Add some more text explaining what this option means (and that is queries an external server).
- [ ] Tests?

### CODE REVIEW

File: templates/settings-admin.php [hunk 1]
```diff
@@ -55,4 +55,9 @@
 			   value="1" <?php if ($_['enforceSpecialCharacters']) print_unescaped('checked="checked"'); ?> />
 		<label for="password-policy-enforce-special-characters"><?php p($l->t('Enforce special characters'));?></label><br/>
 	</p>
+	<p id="enforceHaveIBeenPwned">
+		<input type="checkbox" name="password-policy-enforce-have-i-been-pwned" id="password-policy-enforce-have-i-been-pwned" class="checkbox"
+			   value="1" <?php if ($_['enforceHaveIBeenPwned']) print_unescaped('checked="checked"'); ?> />
+		<label for="password-policy-enforce-have-i-been-pwned"><?php p($l->t('Enforce checking against the list of breached passwords from haveibeenpwnd.com'));?></label><br/>
```
Comment (MorrisJobke) [inline templates/settings-admin.php]: Check password against the list of breached passwords from haveibeenpwnd.com

Comment (MorrisJobke) [reply templates/settings-admin.php]: Add a little text:

```html
<p class="password-policy-settings-hint">
This check creates a hash of the password and sends the first 5 characters of this hash to the haveibeenpwnd.com API to retrieve a list of all hashes that start with those. Then it checks on the Nextcloud instance if the password hash is in the result set.
</p>
```

Then create following style (there is no suitable in the server styles):

```css
.password-policy-settings-hint {
    opacity: 0.7;
    padding-left: 19px;
}
```

**Comment by MorrisJobke** [comment]
> Add some more text explaining what this option means (and that is queries an external server).

See my inline comment.

File: lib/PasswordValidator.php [hunk 2]
```diff
@@ -155,4 +164,42 @@ protected function checkCommonPasswords($password) {
...
+			} catch (\Exception $e) {
+				return;
+			}
+
+			$result = $response->getBody();
+
+			if (strpos($result, $needle)) {
+				$message = 'Password is present in compromised password list. Please chose a different password.';
+				$message_t = $this->l->t(
+					'Password is present in compromised password list. Please chose a different password.'
```
Comment (chk1) [inline lib/PasswordValidator.php]: Spotted a typo here: 
"Please chose" should be "Please choose" (lines 196 & 198)

File: appinfo/info.xml [hunk 3]
```diff
@@ -2,17 +2,20 @@
...
+	<summary>Allows admins to configure a password policy</summary>
 	<description>
 		Allow admin to define certain pre-conditions for password, e.g. enforce a minimum length
 	</description>
-	<licence>AGPL</licence>
-	<author>Bjoern Schiessle</author>
 	<version>1.4.0</version>
+	<licence>agpl</licence>
+	<author>Bjoern Schiessle</author>
+	<namespace>Password_Policy</namespace>
```
Comment (nickvergessen) [inline appinfo/info.xml]: Just for the record, in theory this is not psr compliant :see_no_evil:

File: lib/PasswordValidator.php [hunk 4]
```diff
@@ -155,4 +164,42 @@ protected function checkCommonPasswords($password) {
...
+						'timeout' => 5
+					]
+				);
+			} catch (\Exception $e) {
+				return;
+			}
+
+			$result = $response->getBody();
+
+			if (strpos($result, $needle)) {
```
Comment (nickvergessen) [inline lib/PasswordValidator.php]: ` !== false`

otherwise you don't match when it's the first result?

File: appinfo/info.xml (hunk 3 above)
Comment (rullzer) [reply appinfo/info.xml]: I know. But lets fix that in a different PR ;)

**Comment by MorrisJobke** [comment]
@rullzer I'm fine with backporting to stable13

**Pull request merged**